File: src/thinking-graph.ts
import type { Note, Relationship } from "./workspace-client"

/**
 * The Thinking Graph as the Note detail surface reads it: related lookup,
 * degree, and the candidates a new Relationship may name. Every function here
 * is a projection of committed state, so nothing in the interface has to know
 * that a symmetric pair is stored with its endpoints sorted.
 *
 * The durable rules — canonical ordering, validation, and cascade — belong to
 * the Rust Thinking Graph module. Nothing here writes.
 */

/** The endpoint that is not `noteId`, or nothing when this pair excludes it. */
function otherEndpoint(relationship: Relationship, noteId: string): string | null {
  if (relationship.noteIdA === noteId) return relationship.noteIdB
  if (relationship.noteIdB === noteId) return relationship.noteIdA
  return null
}

/** The Notes this Note is related to, in the order they were related. */
export function relatedNoteIds(relationships: Relationship[], noteId: string): string[] {
  return relationships.flatMap((relationship) => {
    const other = otherEndpoint(relationship, noteId)
    return other === null ? [] : [other]
  })
}

/** How many Notes this Note is related to. */
export function degree(relationships: Relationship[], noteId: string): number {
  return relatedNoteIds(relationships, noteId).length
}

/**
 * Related Notes as durable Notes. An endpoint with no Note is never shown,
 * though storage's cascade means one cannot arrive here.
 */
export function relatedNotes(
  notes: Note[],
  relationships: Relationship[],
  noteId: string,
): Note[] {
  const related = new Set(relatedNoteIds(relationships, noteId))
  return notes.filter((note) => related.has(note.id))
}

/**
 * What the relation editor may offer: Notes in the same Thinking Workspace,
 * never this Note, never one already related, narrowed by the thinker's text.
 */
export function relatableNotes(
  notes: Note[],
  relationships: Relationship[],
  noteId: string,
  query: string,
): Note[] {
  const note = notes.find((candidate) => candidate.id === noteId)
  if (!note) return []
  const excluded = new Set([noteId, ...relatedNoteIds(relationships, noteId)])
  const terms = query.trim().toLowerCase()
  return notes.filter(
    (candidate) =>
      candidate.workspaceId === note.workspaceId &&
      !excluded.has(candidate.id) &&
      (terms === "" ||
        `${candidate.markdown} ${candidate.annotation ?? ""}`.toLowerCase().includes(terms)),
  )
}

File: src-tauri/src/thinking_graph.rs
//! The Thinking Graph: the meaningful associations among Notes in one Thinking
//! Workspace, independent of how they are displayed.
//!
//! Everything here is a pure rule over committed state. Canonical ordering,
//! related lookup, degree, and pair validation live in this one place, so no
//! caller has to know that a symmetric pair is stored with its endpoints
//! sorted.

use serde::Serialize;

use crate::workspace::Note;

/// Who created a Relationship. This slice writes only `Manual`; `Ai` is part of
/// the durable vocabulary now so a later assistance slice adds rows rather than
/// changing what an existing row means.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RelationshipProvenance {
    Manual,
    Ai,
}

impl RelationshipProvenance {
    pub(crate) fn as_str(self) -> &'static str {
        match self {
            Self::Manual => "manual",
            Self::Ai => "ai",
        }
    }

    pub(crate) fn from_str(value: &str) -> Self {
        match value {
            "ai" => Self::Ai,
            _ => Self::Manual,
        }
    }
}

/// A symmetric, untyped association between two distinct Notes in one Thinking
/// Workspace. It carries no direction and no named relation: `note_id_a` sorts
/// before `note_id_b` only so one pair has one row.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Relationship {
    pub(crate) id: String,
    pub(crate) workspace_id: String,
    pub(crate) note_id_a: String,
    pub(crate) note_id_b: String,
    pub(crate) provenance: RelationshipProvenance,
    pub(crate) created_at: String,
}

impl Relationship {
    /// The endpoint that is not `note_id`, or nothing when this Relationship
    /// does not touch that Note. Related lookup and degree are this, repeated.
    pub(crate) fn other_endpoint(&self, note_id: &str) -> Option<&str> {
        if self.note_id_a == note_id {
            Some(&self.note_id_b)
        } else if self.note_id_b == note_id {
            Some(&self.note_id_a)
        } else {
            None
        }
    }
}

/// Why a proposed pair is not a Relationship the Thinking Graph can hold.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum GraphViolation {
    SelfRelationship,
    CrossWorkspace,
    MissingNote,
}

/// The canonical ordering of one unordered pair, so either endpoint order names
/// the same row.
pub(crate) fn canonical_pair<'a>(left: &'a str, right: &'a str) -> (&'a str, &'a str) {
    if left <= right {
        (left, right)
    } else {
        (right, left)
    }
}

/// Whether these two Notes are already related, in either endpoint order.
pub(crate) fn is_related(relationships: &[Relationship], left: &str, right: &str) -> bool {
    relationships
        .iter()
        .any(|relationship| relationship.other_endpoint(left) == Some(right))
}

/// Validates a proposed pair against the committed Notes and returns the
/// Workspace both belong to. Every rejection is decided before any transaction
/// opens, so an invalid pair can never leave partial state behind.
pub(crate) fn relatable_workspace_id(
    notes: &[Note],
    left: &str,
    right: &str,
) -> Result<String, GraphViolation> {
    if left == right {
        return Err(GraphViolation::SelfRelationship);
    }
    let endpoint = |note_id: &str| {
        notes
            .iter()
            .find(|note| note.id() == note_id)
            .ok_or(GraphViolation::MissingNote)
    };
    let left = endpoint(left)?;
    let right = endpoint(right)?;
    if left.workspace_id() != right.workspace_id() {
        return Err(GraphViolation::CrossWorkspace);
    }
    Ok(left.workspace_id().to_owned())
}
